pkg/scheduler/preemption/fairsharing/ordering.go

```diff
@@ -0,0 +1,195 @@
+		prunedCohorts:       sets.New[*cache.CohortSnapshot](),
+	}
+
+	cohort := cq.Parent()
+	for cohort != nil {
+		t.preemptorAncestors.Insert(cohort)
+		cohort = cohort.Parent()
+	}
+
+	for _, candidate := range candidates {
+		t.clusterQueueToTarget[candidate.ClusterQueue] = append(t.clusterQueueToTarget[candidate.ClusterQueue], candidate)
+	}
+
+	return t
+}
+
+func (t *TargetClusterQueueOrdering) Iter() iter.Seq[*TargetClusterQueue] {
+	return func(yield func(v *TargetClusterQueue) bool) {
+		// handle CQ without Cohort case.
+		if !t.preemptorCq.HasParent() {
+			targetCq := &TargetClusterQueue{
+				ordering: t,
+				targetCq: t.preemptorCq,
+			}
+
+			for targetCq.HasWorkload() {
+				if !yield(targetCq) {
+					return
+				}
+			}
+		}
+
+		root := t.preemptorCq.Parent().Root()
+		// we stop once we have marked the root as pruned.
+		for !t.prunedCohorts.Has(root) {
+			targetCq := t.nextTarget(root)
+
+			// an iteration which just pruned some node(s).
+			if targetCq == nil {
+				continue
+			}
+			if !yield(targetCq) {
+				return
+			}
+		}
+	}
+}
+
+// DropQueue indicates that we should no longer
+// consider workloads from this Queue.
+func (t *TargetClusterQueueOrdering) DropQueue(cq *TargetClusterQueue) {
+	t.prunedClusterQueues.Insert(cq.targetCq)
+}
+
+func (t *TargetClusterQueueOrdering) onPathToPreemptorCQ(cohort *cache.CohortSnapshot) bool {
+	return t.preemptorAncestors.Has(cohort)
+}
+
+func (t *TargetClusterQueueOrdering) hasWorkload(cq *cache.ClusterQueueSnapshot) bool {
+	return len(t.clusterQueueToTarget[cq.GetName()]) > 0
+}
+
+// nextTarget is a recursive algorithm for finding the next
+// TargetClusterQueue.  It finds the child with the highest DRS,
+// returning it if it is a ClusterQueue, or entering the recursive
+// call if it is a Cohort.  The return of nil doesn't mean that there
+// are no more candidate ClusterQueues; an iteration may have only
+// pruned nodes from the tree.
+func (t *TargetClusterQueueOrdering) nextTarget(cohort *cache.CohortSnapshot) *TargetClusterQueue {
+	var highestCq *cache.ClusterQueueSnapshot = nil
+	highestCqDrs := -1
+	for _, cq := range cohort.ChildCQs() {
+		if t.prunedClusterQueues.Has(cq) {
+			continue
+		}
+
+		drs := cq.DominantResourceShare()
+		// we can't prune the preemptor ClusterQueue itself,
+		// until it runs out of candidates.
+		if (drs == 0 && cq != t.preemptorCq) || !t.hasWorkload(cq) {
+			t.prunedClusterQueues.Insert(cq)
+		} else if drs >= highestCqDrs {
+			highestCqDrs = drs
+			highestCq = cq
+		}
+	}
+
+	var highestCohort *cache.CohortSnapshot = nil
+	highestCohortDrs := -1
+	for _, cohort := range cohort.ChildCohorts() {
+		if t.prunedCohorts.Has(cohort) {
+			continue
+		}
+
+		drs := cohort.DominantResourceShare()
+
+		// we prune a Cohort when it is no longer borrowing
+		// (DRS=0). Even when not borrowing, we can't prune a
+		// Cohort on path from preemptor ClusterQueue to
+		// root, as there may be imbalance within some
```

Comment:
non-blocking question: if the cohort is not borrowing (DRS=0), but has some internal imbalance, are we in position to reclaim quota from the children. IIUC the picture is like this:

```
       ROOT
   /            \
  L              R
 /  \           /   \
a   b       c     d
```
Now, we are trying to admit into "a", and we see that R has DRS=0, but "d" has DRS > 0. I would expect we can already prune, no?

Reply:
> non-blocking question: if the cohort is not borrowing (DRS=0), but has some internal imbalance, are we in position to reclaim quota from the children. IIUC the picture is like this:

No, we cannot reclaim from any of R's children in this case, since R's DRS=0

> Now, we are trying to admit into "a", and we see that R has DRS=0, but "d" has DRS > 0. I would expect we can already prune, no?

Correct - we will prune R

Reply:
so isn't the comment misleading?

Reply:
This comment is talking about the path from A to root - so L is not pruned just because its DRS is 0 - since there is still potential to preempt from either a or b. It is only pruned once all of its children - a and b - are will no longer yield candidates

Reply:
I see, in that case I would just suggest to rename `onPathToPreemptorCQ` -> `onPathFromRootToPreemptorCQ`. Yes, it is longer, but the "root" seems important to make the function name "readable" without looking into implementation. wdyt?